### ISSUE TITLE
PSyclone losing some subroutines from a subroutine contains statement

### ISSUE THREAD

**Issue by LonelyCat124** [stfc/PSyclone#2205]
I don't have a good understanding of what is happening in this file in Socrates.

File in question is `socrates_set_diag.f90`, which contains a module containing a subroutine containing some additional subroutines.

This could therefore be related directly to #2200 - but in this case PSyclone execution is completed and outputting invalid Fortran code.

Rough code structure is
```
Module socrates_set_diag
....
contains
subroutine set_diag(...)
...
call sum_flux_channels(.., surf=diag%...)
...
contains
subroutine sum_flux_channels(...)
subroutine ... (more than 1)
end subroutine set_diag
end module
```

The output code from Psyclone outputs the subroutine without the `contains` block, so compilation fails as the compiler has no idea what these functions are.

**Comment by rupertford** [comment]
I don't think we've had to deal with code that has subroutines within subroutines before, so you're probably picking up bugs associated with that.

**Comment by LonelyCat124** [comment]
Yeah - it seems quite commonplace in socrates (I'd guess at least 10-15 files may have this behaviour) -  some files will compile even though the contained subroutines are lost, but then fail to link.

**Comment by arporter** [comment]
It's something that happens in NEMOVAR but (AFAIK), we do successfully create CodeBlocks there. There must be something else going on here.

**Comment by LonelyCat124** [comment]
I closed #2200 in favour of this.

From the original issue:
I managed to construct a mini psyclone test that replicates this behaviour:
```
def test_fail_case(fortran_reader):
    code = '''subroutine a(b, c, d)
    real b, c, d

    b = my_func(c, d)

    contains
    real function my_func(a1, a2)
    real a1, a2
    my_func = a1 * a2
    end function
    end subroutine'''
    psyir = fortran_reader.psyir_from_source(code)
```

Test fails here:
```
../../../psyir/frontend/fortran.py:79: in psyir_from_source
    psyir = self._processor.generate_psyir(parse_tree)
../../../psyir/frontend/fparser2.py:1155: in generate_psyir
    self.process_nodes(node, [parse_tree])
../../../psyir/frontend/fparser2.py:2534: in process_nodes
    psy_child = self._create_child(child, parent)
../../../psyir/frontend/fparser2.py:2597: in _create_child
    return handler(child, parent)
../../../psyir/frontend/fparser2.py:4604: in _program_handler
    self.process_nodes(file_container, node.children)
../../../psyir/frontend/fparser2.py:2534: in process_nodes
    psy_child = self._create_child(child, parent)
../../../psyir/frontend/fparser2.py:2597: in _create_child
    return handler(child, parent)
../../../psyir/frontend/fparser2.py:4489: in _subroutine_handler
    self.process_nodes(routine, sub_exec.content)
../../../psyir/frontend/fparser2.py:2534: in process_nodes
    psy_child = self._create_child(child, parent)
../../../psyir/frontend/fparser2.py:2597: in _create_child
    return handler(child, parent)
../../../psyir/frontend/fparser2.py:3662: in _assignment_handler
    self.process_nodes(parent=assignment, nodes=[node.items[2]])
../../../psyir/frontend/fparser2.py:2534: in process_nodes
    psy_child = self._create_child(child, parent)
../../../psyir/frontend/fparser2.py:2597: in _create_child
    return handler(child, parent)
../../../psyir/frontend/fparser2.py:4047: in _part_ref_handler
    symbol = _find_or_create_imported_symbol(parent, reference_name)

psyclone.psyir.symbols.symbol.SymbolError: PSyclone SymbolTable error: No Symbol found for name 'my_func'.
```

**Comment by LonelyCat124** [comment]
They definitely seem related, and I'm not sure what is special here.
I tried this test:
```
    def test_fail_case2(fortran_reader, fortran_writer):
        code = '''subroutine a(b, c, d)
        real b, c, d

        b = 1.0

        contains
        real function my_func(a1, a2)
        real a1, a2
        my_func = a1 * a2
        end function
        end subroutine'''
        psyir = fortran_reader.psyir_from_source(code)
        print(fortran_writer(psyir))
       assert False
```

and the output is:
```
subroutine a(b, c, d)
  real :: b
  real :: c
  real :: d

  b = 1.0

end subroutine a
```

So the subfunction dissapears.

The subroutine is not converted to a CodeBlock - i have a `Routine[name:'a']` as the first child of `psyir`

I worked around it by adding:
```
        try:
            subprogram_part = _first_type_match(node.children, Fortran2003.Internal_Subprogram_Part)
            has_contains = True
        except ValueError:
            has_contains = False
        if has_contains:
            raise NotImplementedError("PSyclone doesn't yet support contains inside a Subroutine or Function")
```
to the start of `_subroutine_handler` in `fparser.py`, as the first code executed. This allows this test to work and results in a codeblock, however I don't know if we want to support this outside of the CodeBlock solution? @arporter perhaps we use this solution as a temporary measure and add full support later?

**Comment by arporter** [comment]
I've searched through `psyir/frontend/fparser2.py` and we simply don't mention this problem anywhere - we catch CONTAINS within structure definitions and handle the CONTAINS for MODULEs but that's it. Your workaround would be exactly what I would do as a first step - I'd suggest doing that as an initial PR so that we at least don't crash.

We need a design discussion about how/if we handle this feature of Fortran in the longer term.

**Comment by sergisiso** [comment]
The problem is where to put it? Currently Routines are Schedules. We could put it there but, a walk for a node type would find the internal subroutine nodes, is that what we want? Also dependency forward backward references would need special cases. And similar issues.

Another possibility is for Routines to have two children, body and subroutines. But would require another layer and extra .parent/.children[0] (or semantic version) in many places in the code.

I don't know what's better.

**Comment by rupertford** [comment]
A PSyIR solution should solve both @LonelyCat124, as programs and subroutine's are both Routine's in PSyIR.

I would go with Andy's suggestion (mostly). If it looks and smell like a Routine, it is a Routine. We should be able to re-use most of the handlers. Where I think the issues are are:

It isn't a Routine in the PSyIR sense as it has different properties a) it must be contained within a Routine, b) it must use the Routine's symbol table (I believe) and c) the rules about what can be a child of what in a Routine would have to change.

We therefore need to be able to distinguish between this routine - (InternalRoutine?) - and a general Routine. We can natually do this in two ways 1) subclass to e.g. InternalRoutine or 2) have as is_internal flag as we have is_program for program.

The using of a parent's symbol table could be interesting coding, but hopefully it could just be a constraint when constructing the PSyIR and a search through parent symbol tables would just jump to the correct one.

No matter what, we are going to have to change the assumptions about Routine's in our internal search's etc. but there is no way round that when we introduce a new concept. We just need to make sure we can distinguish when we need to.

Happy to be shot down :-)

**Comment by arporter** [comment]
I don't think the SymbolTable thing would be a biggy - it should 'just work' in the same way that a subroutine has access to the symbols defined in its parent module.

I was going to say that n alternative solution would be to 'canonicalise' the PSyIR and move any 'internal' routines up into the parent module. However, given that they share the scope of the program unit they are contained in, I can't actually see any nice way of doing that.

Another potential solution (possibly the best?) would be to inline all calls to the 'internal' routine.

**Comment by rupertford** [comment]
> I don't think the SymbolTable thing would be a biggy - it should 'just work' in the same way that a subroutine has access to the symbols defined in its parent module.

I think the InternalRoutine's symbol table should be the same symbol table as its parent Routine. Hopefully as you say it would not be a biggy.

> I was going to say that n alternative solution would be to 'canonicalise' the PSyIR and move any 'internal' routines up into the parent module. However, given that they share the scope of the program unit they are contained in, I can't actually see any nice way of doing that.

Aren't we talking about cases where there may be no `module` i.e. just a program or just a subroutine. At least we have to support these cases. Or do you mean something else?

> Another potential solution (possibly the best?) would be to inline all calls to the 'internal' routine.

I think this should be an optimisation, but it is a clever (sneaky) idea  :-)

**Comment by arporter** [comment]
> > I don't think the SymbolTable thing would be a biggy - it should 'just work' in the same way that a subroutine has access to the symbols defined in its parent module.
> 
> I think the InternalRoutine's symbol table should be the same symbol table as its parent Routine. Hopefully as you say it would not be a biggy.
> 
> >

I don't think that would work. Any symbols declared locally in the 'internal' routine are not visible in the 'parent' routine.

**Comment by LonelyCat124** [comment]
Could the symbol table be some sort of proxy class, that both is a symbol table but when doing things like lookup also searches in the parent routines symbol table if its not found locally?I think we already do something similar for scoping somewhere?

**Comment by LonelyCat124** [comment]
I think these things are still complicated from Sergi's point though, if these routines are not `children` and thus are hidden from things like `psyir.walk` then we might miss things, but if they are it might flag false positives I think.

**Comment by rupertford** [comment]
> > > I don't think the SymbolTable thing would be a biggy - it should 'just work' in the same way that a subroutine has access to the symbols defined in its parent module.
> > 
> > 
> > I think the InternalRoutine's symbol table should be the same symbol table as its parent Routine. Hopefully as you say it would not be a biggy.
> > >
> 
> I don't think that would work. Any symbols declared locally in the 'internal' routine are not visible in the 'parent' routine.

Oops yes, my bad. We would need a local symbol table. If anything that is easier I think.

**Comment by arporter** [comment]
I like Sergi's suggestion of letting `Routine` have two children where the second would be a list of 'internal' routines. Since we don't typically investigate the contents of a `Routine` from outside of it (unless inlining) I think this would avoid the accidental walk of nodes we didn't want and, on the rare occasions that we care, we can still find them (albeit, we'll have to modify all such places in the code as Sergi says).

**Comment by rupertford** [comment]
> I think these things are still complicated from Sergi's point though, if these routines are not `children` and thus are hidden from things like `psyir.walk` then we might miss things, but if they are it might flag false positives I think.

I would expect to use children so a standard walk and a walk with Routine, would (and should probably) find the nodes but this is a new concept so we would have to fix our existing implementation or modify the search (e.g. by adding a not X option, or only_this_type option so we could say not IntrinsicRoutine, or only_this_type Routine if we do sublcass). I don't see a way round it and am not sure we should be looking for one.

**Comment by rupertford** [comment]
I guess they could be hidden as a list, but we don't do that for Routines in modules, for example. Would be not be treating these differently to all other nodes, even though they are essentially a first class concept, to make life easier for ourselves in the current implementation given that this is a new concept in PSyIR?

**Comment by sergisiso** [comment]
Note that I was not proposing to hide anything. I was proposing to have the schedule of statements be a children 0 instead of inheriting from Schedule, and give it a semantic name e.g body, so

routine.body.walk(Reference) would return only the refernces in the execution statement
routine.subcontainer.walk(Reference) would return only the refernces in the sub-subroutines
routine.walk(Reference) wourld return all references

like the rest of the PSyIR works.

**Comment by rupertford** [comment]
If that is the consensus and it minimises changes then fine. We would need to find/walk routines first then look to see if there are any internalroutines (I don't think they are subcontainers). Presumably the names of these internalroutines can be stored for the backend?
